sourcepolicy/engine.go

```diff
@@ -0,0 +1,199 @@
+	}
+	if first {
+		ctx = bklog.WithLogger(ctx, bklog.G(ctx).WithField("orig", *srcOp).WithField("updated", op.GetSource()))
+		first = false
+	}
+
+	ident := srcOp.GetIdentifier()
+	scheme, ref, found := strings.Cut(ident, "://")
+	if !found || ref == "" {
+		return mutated, errors.Errorf("failed to parse %q", ident)
+	}
+
+	ctx = bklog.WithLogger(ctx, bklog.G(ctx).WithFields(map[string]interface{}{
+		"scheme": scheme,
+		"ref":    ref,
+	}))
+
+	for _, rule := range e.pol.Rules {
+		mut, err := e.evaluateRule(ctx, rule, scheme, ref, srcOp, &st)
+		if mut {
+			bklog.G(ctx).Debug("sourcepolicy: mutated source")
+			mutated = true
+		}
+
+		if err != nil {
+			return mutated, err
+		}
+		if mut {
+			// Re-evalutate ruleset
+			count++
+			if count > maxIterr {
+				return mutated, errors.Wrap(ErrTooManyOps, ident)
+			}
+			goto Eval
+		}
+	}
+
+	return mutated, nil
+}
+
+// temporary state while evaluating a policy
+type evalState struct {
+	allowed map[string]bool
+	regexes map[string]*regexp.Regexp
+}
+
+func (st *evalState) allow(ref string) {
+	if st.allowed == nil {
+		st.allowed = map[string]bool{}
+	}
+	st.allowed[ref] = true
+}
+
+func (st *evalState) regex(key, pattern string) (*regexp.Regexp, error) {
+	if st.regexes == nil {
+		st.regexes = map[string]*regexp.Regexp{}
+	}
+	if r, ok := st.regexes[key]; ok {
+		return r, nil
+	}
+	r, err := regexp.Compile(pattern)
+	if err != nil {
+		return nil, err
+	}
+	st.regexes[key] = r
+	return r, nil
+}
+
+func (e *Engine) evaluateRule(ctx context.Context, rule *spb.Rule, scheme, ref string, op *pb.SourceOp, st *evalState) (bool, error) {
+	switch scheme {
+	case "http", "https":
+		// The scheme ref is important for http/https sources
+		ref = scheme + "://" + ref
+
+		// Update the scheme to match the rule
+		// This is done so the rule can match regardless of what shceme we pulled off the URL and the rule can be written with either scheme.
+		switch rule.Source.Type {
+		case "http", "https":
+			scheme = rule.Source.Type
+		}
+	}
+
+	var regex *regexp.Regexp
+	switch rule.Source.MatchType {
+	case spb.MatchType_EXACT:
+	case spb.MatchType_REGEX:
+		var err error
+		regex, err = st.regex(rule.Source.Type+"://"+rule.Source.Identifier, rule.Source.Identifier)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	match, err := e.matcher.Match(ctx, rule.Source, scheme, ref, op.Attrs, regex)
+	if err != nil {
+		return false, errors.Wrap(err, "error evaluating rule")
+	}
+
+	bklog.G(ctx).Debug("sourcepolicy: rule match")
+
+	if !match {
+		return false, nil
+	}
+
+	switch rule.Action {
+	case spb.PolicyAction_ALLOW:
+		st.allow(ref)
+		return false, nil
+	case spb.PolicyAction_DENY:
+		// If this has already been allowed by a previous rule then we can ignore it.
+		if st.allowed[ref] {
+			return false, nil
+		}
+		if match {
+			return false, errors.Wrapf(ErrSourceDenied, "rule %s %s applies to source %s://%s", rule.Action, rule.Source.Identifier, scheme, ref)
+		}
+		return false, nil
```

Comment:
So it looks like this seems to implement a "first-matching rule applies" here. So, if `ALLOW` is set for a ref, it can't be undone later in the policy. Is there a good reason to go with this over a "last-matching rule applies" (so that `ALLOW`s can override earlier `DENY`s and vice-versa).

I think they're both equally powerful, I'm just curious if there's a convention we should stick to, or one that provides a more intuitive user experience. I don't think I have a preference, but kinda curious about this.

Reply:
A bit hairy but maybe we can let user control this with an `Order` directive that could default to `DENY,ALLOW`. If a source is excluded by DENY, it is excluded unless it matches ALLOW. If neither is matched, the source is allowed. If user wants the other way around they can set `Order` to `ALLOW,DENY`.

Reply:
I think rule precedence is my biggest concern on this and is probably the most hairy thing in terms of user expectation, but maybe we can control that a couple of different ways:

1. Versioning
2. Plug-ability

On 2, I think it could be possible to have something like run an image to check policies.

Reply:
Hm, I agree that changing the `Order` sounds a bit hairy - if we do decide we want that, I think pluggability would be the way to go instead of extending the builtin.

For now, I think we can just select either order, as long as we can justify it and compare to existing tools (will also help with explaining to users in docs etc).

As I can see it, we have 3 options:
- First-matching `ALLOW`/`DENY` rule applies (currently implemented)
  - I think this might be confusing, since `CONVERT` becomes a special case.
- First-matching `ALLOW`/`DENY`/`CONVERT` rule applies (same as above +`CONVERT`)
  - This seems equivalent to the first, but applies the same consistency to early-exit on *any* matching rule. This means that most `CONVERT` rules would need to be near the top of the list, but, I think this matches up with user-expectations well, I think this would be most similar to iptables and similar?
- Last-matching `ALLOW`/`DENY` rule applies (described in https://github.com/moby/buildkit/pull/2943#discussion_r949748366)

Reply:
https://httpd.apache.org/docs/2.2/mod/mod_authz_host.html#order might be worth to look at.

Reply:
Not sure I follow if we need to make a change here.
Does that mean if there is an allow then a deny it should deny it?
Likewise for deny then allow, it should allow it?

Reply:
> Does that mean if there is an allow then a deny it should deny it?

Yes. Eg. `ALLOW busybox:*; DENY busybox:1.2` means `busybox:1.2` is not allowed.

I don't think the map is `st` is needed.

Reply:
Should this be based on precision of the match or just blanket that a deny regardless of precision is always applied?
E.g. `ALLOW busybox:latest; DENY busybox:*`?

Reply:
Also another example, what about `ALLOW busybox:latest; DENY busybox:latest`

Reply:
IMO, the final rule should apply. 
- `ALLOW busybox:latest; DENY busybox:*` = `busybox:*` is denied (including `busybox:latest`).
- `ALLOW busybox:latest; DENY busybox:latest` = `busybox:latest` is denied

Reply:
Agree, just looking at the order is simple. Judging what rule is more precise than another one will get messy quickly.

Reply:
Done.